content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,157 @@
+
+Context fetching for autocomplete relies mostly on fast classical and compiler-based techniques, rather than LLMs, because we need something that performs very quickly and can be invoked at high volume. As long as the quality of the selected snippets is good enough, the LLM model can use that information to produce high-quality suggestions.
+
+---
+
+## Context fetching for test generation
+
+This feature allows you to select a file and automatically generate tests for it. The requirements for this feature are vastly different from those for chat or autocomplete. For this, we want to be able to detect if there is already a test file for the chosen file. If it doesn’t exist, we want to generate that automatically. Whether you end up generating a new file or finding an existing test file for the chosen file, we want to be able to add relevant files for context.
+
+![16_test](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/16_test.png)
+
+To achieve this, after creating or finding the test file, we find other test files to understand how tests are written in the given project. This ensures that the AI doesn’t hallucinate and uses the appropriate methods and conventions. Without this context, test generation wouldn’t work as well because it might use outdated methods or methods not used in the given project.
+
+![17_test_context](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/17_test_context.png)
+
+---
+
+## Context fetching for editing code
+
+This is a feature that lets you select a block of code and edit it using a natural language instruction.
+
+![18_edits](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/18_edits.png)
+For this, we automatically retrieve:
+
+- Surrounding code, both before and after the cursor
+- Diagnostic information like warnings and errors
+- User-specified context from @-mentions
+
+By including diagnostic information, we're able to provide more appropriate code edits to the selected range of code. In the future, we plan to incorporate code graph context here as well.
+
+---
+
+## Further room for improvement
+
+Even though creating and polishing a context fetching mechanism for each of these features has taken a lot of work, we must admit that there’s still a lot of room for improvement.
+
+Some of the areas we’re still actively working are:
+
+- Underlying LLM models, comparing different models and [fine-tuning](https://sourcegraph.com/blog/enhancing-code-completion-for-rust-in-cody) some of them
+- A query rewriting strategy where we have AI rewrite user queries to fit our context-fetching system better
+- (Work in progress) a smarter way to determine what types of context need to be fetched for each query
```

Comment:
```suggestion
- A smarter way to determine what types of context need to be fetched for each query
```

we already mention before we get into the list that these are things we're actively working on so I don't think it's necessary to call it out again.